From 7690ca4f793771009efe7355fc7cc449b29405ab Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Thu, 7 Apr 2022 12:39:07 +0200
Subject: [PATCH] Integrate with Balena u-boot environment

Integrate tx2 on L4T 32.4.2 with Balena u-boot env

This applies to the TX2, TX2 4GB and TX2 NX

Upstream-Status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 configs/p2771-0000-500_defconfig    | 23 +++++++++++++--------
 configs/p3636-0001_defconfig        | 13 +++++++++---
 include/config_distro_bootcmd.h     | 31 +++++++++++++++++++++++++++--
 include/configs/p2771-0000.h        |  3 ---
 include/configs/tegra-common-post.h |  5 +----
 5 files changed, 55 insertions(+), 20 deletions(-)

diff --git a/configs/p2771-0000-500_defconfig b/configs/p2771-0000-500_defconfig
index 1ff344240f..26e59c45b1 100644
--- a/configs/p2771-0000-500_defconfig
+++ b/configs/p2771-0000-500_defconfig
@@ -1,7 +1,7 @@
 CONFIG_ARM=y
 CONFIG_TEGRA=y
 CONFIG_SYS_TEXT_BASE=0x80080000
-CONFIG_ENV_SIZE=0x2000
+CONFIG_ENV_SIZE=0x3000
 CONFIG_ENV_OFFSET=0xFFFD8000
 CONFIG_NR_DRAM_BANKS=1026
 CONFIG_TEGRA186=y
@@ -14,10 +14,10 @@ CONFIG_SYS_PROMPT="Tegra186 (P2771-0000-500) # "
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
-CONFIG_CMD_PCI=y
+#CONFIG_CMD_PCI=y
 CONFIG_CMD_SPI=y
-CONFIG_CMD_USB=y
-# CONFIG_CMD_SETEXPR is not set
+#CONFIG_CMD_USB=y
+CONFIG_CMD_SETEXPR=y
 # CONFIG_CMD_NFS is not set
 CONFIG_CMD_EXT4_WRITE=y
 CONFIG_DEFAULT_DEVICE_TREE="tegra186-p2771-0000-500"
@@ -34,10 +34,17 @@ CONFIG_PCI_TEGRA=y
 CONFIG_POWER_DOMAIN=y
 CONFIG_TEGRA186_POWER_DOMAIN=y
 CONFIG_SYS_NS16550=y
-CONFIG_USB=y
-CONFIG_DM_USB=y
+#CONFIG_USB=y
+#CONFIG_DM_USB=y
 CONFIG_POSITION_INDEPENDENT=y
-CONFIG_BOOTP_PREFER_SERVERIP=y
+#CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_NVME=y
+#CONFIG_NVME=y
 CONFIG_OF_LIBFDT_OVERLAY=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_CMD_FAT=y
+CONFIG_ENV_IS_NOWHERE=y
+CONFIG_CMD_FS_UUID=y
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_EDITENV=y
diff --git a/configs/p3636-0001_defconfig b/configs/p3636-0001_defconfig
index 9117982fb9..3763ec3fb3 100644
--- a/configs/p3636-0001_defconfig
+++ b/configs/p3636-0001_defconfig
@@ -1,7 +1,7 @@
 CONFIG_ARM=y
 CONFIG_TEGRA=y
 CONFIG_SYS_TEXT_BASE=0x80080000
-CONFIG_ENV_SIZE=0x2000
+CONFIG_ENV_SIZE=0x3000
 CONFIG_ENV_OFFSET=0xFFFD8000
 CONFIG_NR_DRAM_BANKS=1026
 CONFIG_TEGRA186=y
@@ -18,7 +18,7 @@ CONFIG_CMD_MMC=y
 CONFIG_CMD_PCI=y
 CONFIG_CMD_SPI=y
 CONFIG_CMD_USB=y
-# CONFIG_CMD_SETEXPR is not set
+CONFIG_CMD_SETEXPR=y
 # CONFIG_CMD_NFS is not set
 CONFIG_CMD_EXT4_WRITE=y
 CONFIG_DEFAULT_DEVICE_TREE="tegra186-p3636-0001"
@@ -40,5 +40,12 @@ CONFIG_DM_USB=y
 CONFIG_POSITION_INDEPENDENT=y
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_NVME=y
+#CONFIG_NVME=y
 CONFIG_OF_LIBFDT_OVERLAY=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_CMD_FAT=y
+CONFIG_ENV_IS_NOWHERE=y
+CONFIG_CMD_FS_UUID=y
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_EDITENV=y
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index 599e2b7212..f9aa73958a 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -25,6 +25,7 @@
  * message that includes some other pre-processor symbols in the text.
  */
 
+#define BOOT_TARGET_DEFAULTDEVPLIST    "18"
 #define BOOTENV_SHARED_BLKDEV_BODY(devtypel) \
 		"if " #devtypel " dev ${devnum}; then " \
 			"devtype=" #devtypel "; " \
@@ -429,6 +430,7 @@
 	BOOTENV_SHARED_UBIFS \
 	BOOTENV_SHARED_EFI \
 	BOOTENV_SHARED_VIRTIO \
+	"board_fdt_file=" __stringify(TEGRA_BOARD_FDT_FILE) "\0" \
 	"boot_prefixes=/ /boot/\0" \
 	"boot_scripts=boot.scr.uimg boot.scr\0" \
 	"boot_script_dhcp=boot.scr.uimg\0" \
@@ -449,7 +451,7 @@
 		"fi\0"                                                    \
 	\
 	"boot_a_script="                                                  \
-		"load ${devtype} ${devnum}:${distro_bootpart} "           \
+		"load ${resin_dev_type} ${resin_dev_index}:${resin_boot_part} "           \
 			"${scriptaddr} ${prefix}${script}; "              \
 		"source ${scriptaddr}\0"                                  \
 	\
@@ -465,6 +467,14 @@
 			"fi; "                                            \
 		"done\0"                                                  \
 	\
+        "resin_set_boot_part_tx2=part index mmc 0 resin-boot"             \
+                                " resin_boot_part; echo New boot part:"   \
+                                " $resin_boot_part;\0"                    \
+        \
+        "resin_update_boot_part_tx2=if test \"$target\" = \"mmc0\"; then" \
+                                 " echo Old boot part: $resin_boot_part;" \
+                                 " run resin_set_boot_part_tx2; fi\0"     \
+        \
 	"scan_dev_for_boot="                                              \
 		"echo Scanning ${devtype} "                               \
 				"${devnum}:${distro_bootpart}...; "       \
@@ -476,7 +486,18 @@
 		"\0"                                                      \
 	\
 	"scan_dev_for_boot_part="                                         \
-		"part list ${devtype} ${devnum} -bootable devplist; "     \
+		"if test \"0\" =  \"${devnum}\"; "                        \
+		"then "                                                   \
+		"   if test -e ${devtype} ${devnum}:${resin_root_part} boot/extlinux/extlinux.conf; then " \
+		"       echo Found extlinux.conf on root part: ${resin_root_part}; " \
+		"       setenv devplist ${resin_root_part}; "             \
+		"   else "                                                \
+		"       echo Could not find extlinux.conf on root part, using boot part: ${resin_boot_part}; " \
+		"       setenv devplist ${resin_boot_part}; "             \
+		"   fi; "                                                 \
+		"else "                                                   \
+		"   part list ${devtype} ${devnum} -bootable devplist; "  \
+		"fi; "							  \
 		"env exists devplist || setenv devplist 1; "              \
 		"for distro_bootpart in ${devplist}; do "                 \
 			"if fstype ${devtype} "                           \
@@ -493,8 +514,14 @@
 		BOOTENV_SET_NVME_NEED_INIT                                \
 		BOOTENV_SET_IDE_NEED_INIT                                 \
 		BOOTENV_SET_VIRTIO_NEED_INIT                              \
+		"setenv resin_kernel_load_addr ${loadaddr}; "                    \
 		"for target in ${boot_targets}; do "                      \
+                    "if env exist resin_set_kernel_root; then "           \
+                        "run resin_update_boot_part_tx2; "                \
+                        "run resin_set_kernel_root; "                     \
+                        "run set_os_cmdline; "                            \
 			"run bootcmd_${target}; "                         \
+                    "fi; "                     \
 		"done\0"
 
 #ifndef CONFIG_BOOTCOMMAND
diff --git a/include/configs/p2771-0000.h b/include/configs/p2771-0000.h
index 30638f3a7e..76e46cf344 100644
--- a/include/configs/p2771-0000.h
+++ b/include/configs/p2771-0000.h
@@ -21,10 +21,7 @@
 #define BOOT_TARGET_DEVICES(func) \
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 0) \
-	func(USB, usb, 0) \
-	func(NVME, nvme, 0) \
 	func(PXE, pxe, na) \
-	func(DHCP, dhcp, na)
 
 #define BOARD_EXTRA_ENV_SETTINGS \
 	"calculated_vars=kernel_addr_r fdt_addr_r scriptaddr pxefile_addr_r " \
diff --git a/include/configs/tegra-common-post.h b/include/configs/tegra-common-post.h
index 6b874c810b..4fefbcf6fc 100644
--- a/include/configs/tegra-common-post.h
+++ b/include/configs/tegra-common-post.h
@@ -25,9 +25,7 @@
 #define BOOT_TARGET_DEVICES(func) \
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 0) \
-	func(USB, usb, 0) \
-	func(PXE, pxe, na) \
-	func(DHCP, dhcp, na)
+	func(PXE, pxe, na)
 #endif
 #include <config_distro_bootcmd.h>
 #else
@@ -98,7 +96,6 @@
 	"fdt_high=" FDT_HIGH "\0" \
 	"initrd_high=" INITRD_HIGH "\0" \
 	BOOTENV \
-	ROOTFS_AB_SELECT \
 	BOARD_EXTRA_ENV_SETTINGS \
 	CONFIG_CHROMEOS_EXTRA_ENV_SETTINGS
 
-- 
2.17.1

